doc/design_docs/packed_conditional_fields.md

```diff
@@ -0,0 +1,205 @@
+###### Example 3. Conditional alignment-packed
+```
+struct ConditionalWithAlignment:
+  0                [+2]  UInt    header_length (h)
+  $align($next, 4) [+h]  Header  header
+  if !header.basic_flag:
+    $align($next, 4) [+2]  UInt    body_length
+  $align($next, 4) [+b]  Body    body
+  $align($next, 4) [+4]  UInt    crc
+  let b = $present(body_length) ? body_length : 16
+```
+
+#### Conditional Block Improvements
+
+Currently Emboss does not support nested conditional blocks, nor does it
+support `else` or `elif` conditional blocks. That said, these are mentioned in
+the Emboss code comments as potential future features and are common in
+languages that offer conditionals, so this proposal considers how they may
+impact the new `$next` behavior.
+
+## Proposed Implementation
+
+A high level description of the proposed algorithm for implementing the new
+behavior in `$next` (or a new symbol chosen for backward-compatibility) is
+provided below, with some error and edge condition checking omitted for
+brevity:
+
+ - Similar to the existing algorithm, traverse the IR for `Structure`s
+   (`struct`s and `bit`s), but for each structure encountered:
+   - Initialize the `lookback` to an empty stack.
+   - Initialize `last` to `None`.
+   - Iterate over expressions (looking for `$next`):
+     - As an incidental action, for each non-virtual and non-synthetic field
+       encountered:
+       - If `last.existence_condition` is constant and
+         `last.existence_condition` is `true`:
+         - Clear `lookback` entirely.
+         - Push `last` onto `lookback`.
+       - Otherwise, if `last.existence_condition` is not the same expression
+         as the current `existence_condition`:
+         - If any fields in `lookback` have the same `existence_condition` as
+           the current field, pop all fields up to and including that field.
+         - Push `last` onto `lookback`.
+       - Set `last` to the current field.
+     - For each `$next` encountered:
+       - Initialize `replacement` to be `None`.
+       - For each reference in `lookback` as `field`, starting from the bottom
+         of the stack and iterating towards the top:
+         - If `replacement` is `None`
+           - Set `replacement` to `field.start_location + field.size`
+           - Note: This branch is equivalent to the existing implementation.
+         - Otherwise:
+           - Copy `$present(f) ? next : previous` as an expression skeleton
+           - Replace
+             - `f` with a `FieldRerference` to `field`
+             - `next` with `field.start_location + field.size`.
+             - `previous` with `replacement`.
+           - Set `replacement` to this updated expression.
+       - Replace `$next` with `replacement`.
+
+The purpose of using a stack for `lookback` is twofold, firstly it ensures
+correct operation with chained conditional blocks one-after-another, such as in
+[Example 4](#example-4-consecutive-conditional-blocks), as well as future
```

Comment:
The `lookback` stack and some of the condition checking are optimizations that, I think, make this section harder to understand.

I think the new (unoptimized) implementation of `$next` would just be:

1.  For each structure:
    1.  Initialize `next` to `0`.
    2.  For each physical field:
        1.  Replace `$next` in appropriate contexts with the current value of `next`.
        2.  Update the definition of `next` to `field_condition ? field_start + field_size : (previous_next)`.

It may even be better to use the unoptimized version in the desugaring stage, then let a later optimization pass strip out the unreachable branches -- I have some related code that I was working on recently that is intended to optimize the `$size_in_bytes` expression, which has many of the same problems.

Reply:
That's fair, you're right that is an optimization now that I think about it, and the `next` value can be propagated forward through non-present fields. I'll update this to make it simpler.

I still think there would be value to doing the optimization in the actual code though, I know that every C++ compiler out there would optimize away chained conditionals with constant values or repeated conditions (with only non-volatile and non-aliased components), but it wouldn't be optimized for something like a potential future Python backend, for instance, which would make accessing the `n`th field of a struct an `O(n)` operation, as it would process the conditionals for all `n-1` fields before it to determine the start location, even if said location is statically known.

Reply:
Yeah, it's probably best for some of the optimization to happen inside Emboss.  I'm just thinking it would be better in a separate optimization stage, instead of trying to implement it in the desugaring stage -- separation of concerns, and it would let any similar constructs get optimized.

Reply:
Ah got it, as in a separate optimization pass that resolves conditionals, then it's more general and applies to other designating like `$size_in_bytes` too?

You're right I like that idea a lot. I'll consider it out of scope of this proposal since it won't matter for any decent C++ compiler and we don't currently support any other backend, but I see now that we don't need to preemptively optimize in the desugaring step.

Reply:
Yeah.  The two specific optimizations here are:

*   If a `CHOICE` has a constant condition, replace it with the appropriate branch; i.e. `true ? y : z` => `y` and `false ? : y : z` => `z`.
*   For any expression `x ? y : (x ? z : q)`, replace the inner `(x ? z : q)` with `q`.  There is a larger possible optimization around propagating constraints into choice operator branches, but this would cover a lot of cases and would be fairly easy to implement.

I'll see about getting the prototype optimizer I hacked together onto GitHub for reference.

Reply:
Actually now that I took a second to look through it I think the former is already handled by `constant_value` in `ir_utils`.

Reply:
It's only partially handled by `constant_value` -- that works for `true ? constant : variable` and `false ? variable : constant`, but not for `true ? variable : variable2` or `false ? variable : variable2`, so it would cover:

```
struct Foo:
  0     [+4]  UInt  x
  $next [+4]  UInt y
  $next [+4]  UInt z
```

But not:

```
struct Foo:
  0     [+4]  UInt      x
  $next [+4*x]  UInt:32[] y
  $next [+4]  UInt      z
```